3box integration 2.0

danjm opened the pull request:
This PR adds on to the current 3box-integration PR. Changes include:

- switch to use spaces api instead of profile api
- ensures that state is only restored on import, and only after the user confirms via a modal prompt

e2e tests are in development, but will be added in a future PR

Demo here: https://streamable.com/ovvbf

danjm committed:
Adds threebox controller

danjm committed:
Adds threebox approval modal

danjm committed:
Fix unit tests and lint after addition of threebox

danjm committed:
Correct threebox behaviour after rejecting request for backup; fixes e2e tests.

danjm committed:
Update threebox controller for automatic syncing

danjm committed:
Ensure frontend locale updates when preferences are changed via direct update within controller

danjm committed:
Add toggle in settings for 3box syncing

danjm committed:
Update threebox controller for latest 3box version

danjm committed:
Delete unnecessary frontend changes for threebox integration

danjm committed:
Backing up address book contacts with threebox

danjm committed:
Update unit tests for 3box-integration additions

danjm committed:
Only enable threebox by default for new wallets

danjm committed:
Mock globals for correct unit tests

danjm committed:
3box '1.10.2' -> '^1.10.2'

danjm committed:
Correct capilalization on 3Box

danjm committed:
Use log.debug instead of console.log in threebox controller

danjm committed:
Update yarn.lock

whymarrh committed:
Remove edge build

whymarrh committed:
Split 3box module into background deps js file

whymarrh committed:
extra bundle opts for bg-libs

whymarrh committed:
sync yarn.lock

whymarrh committed:
new3Box logic

danjm committed:
Show confirm threebox restore after import

danjm committed:
Remove bg-libs.js from manifest file for dev builds

danjm committed:
Switch 3Box controller to using the spaces api (instead of the profile api)

danjm committed:
Finalize switching to spaces api and only restoring from 3box after import

danjm committed:
Update metamask-controller-test.js for threebox controller changes

bdresser commented:
This looks great! Small things:
- I think we need to update the button styles to mtach the design system:

<img width="200" alt="Screen Shot 2019-08-07 at 3 11 51 PM" src="https://user-images.githubusercontent.com/1016190/62661511-23153900-b926-11e9-9929-158df2e1c858.png">

<img width="200" alt="Screen Shot 2019-08-07 at 3 12 27 PM" src="https://user-images.githubusercontent.com/1016190/62661518-27415680-b926-11e9-9ea0-085de0182d7b.png">

- Could primary button say "Restore"

danjm committed:
Make threebox modal style consistent with others and update success button wording

danjm commented:
@bdresser 

I made the threebox modal consistent with other modals while updating the wording to restore:

![Screenshot from 2019-08-13 01-46-01](https://user-images.githubusercontent.com/7499938/62914773-5fa0c480-bd6c-11e9-9c30-086c4300db17.png)
![Screenshot from 2019-08-13 01-20-44](https://user-images.githubusercontent.com/7499938/62914775-5fa0c480-bd6c-11e9-85d5-fabc4f437c27.png)
![Screenshot from 2019-08-13 01-20-25](https://user-images.githubusercontent.com/7499938/62914776-5fa0c480-bd6c-11e9-865f-49902497d47c.png)

All of these modals have different buttons from those that you shared a screen shot of above. If the design system specifies a change to these modals design, we can do them all at once in a separate issue.

danjm committed:
Use mock 3box when in test

bdresser commented:
cool, filed [here](https://github.com/MetaMask/metamask-extension/issues/7003), can handle later.

where did the `?` come from in the first line? let's cut it please

<img width="491" alt="Screen Shot 2019-08-13 at 9 04 32 AM" src="https://user-images.githubusercontent.com/1016190/62957309-67328e80-bda9-11e9-807e-112e6e72346f.png">

danjm committed:
Correct 3box modal header

danjm commented:
> where did the ? come from in the first line? let's cut it please

Cut with 8363564f9

I my first job a senior engineer told me that copy+paste is the devil. Its times like these I believe it...

Gudahtt commented on app/scripts/controllers/threebox.js:
```diff
@@ -0,0 +1,181 @@
+    this.addressBookController = addressBookController
+    this.keyringController = keyringController
+    this.provider = this._createProvider({
+      static: {
```
What is this property for? It looks as though it's being ignored in `createMetamaskMiddleware`

danjm replied:
Yep, it can be removed

Gudahtt commented on app/scripts/controllers/threebox.js:
```diff
@@ -0,0 +1,181 @@
+
+      try {
+        this.box = await Box.openBox(address, this.provider)
+        await this._waitForOnSyncDone()
```
What would happen if the network was interrupted during this sync I wonder 🤔 

I thought at first that maybe there was an error not being handled somewhere, but... their API doesn't seem to handle this scenario. I guess the only option is to wait, and handle the case where it times out? 🤷‍♂

danjm replied:
That's a good point. I guess, in the case of a timeout, we should just try again?

Gudahtt replied:
Sure, that could work. Or just throw an error. Or maybe retry, then throw an error after X failures.

danjm replied:
- [ ] do something about this

Gudahtt replied:
After looking over this with @whymarrh , we discovered that the `onSyncDone` function was recently updated to return a promise that correctly handles errors and resolves when the sync is finished. We'll need to update 3Box, but after that we can just await `this.box.onSyncDone()` and it'll "just work". Mostly (the timeout case is still problematic).

The "onSyncDone" from the spaces API is kinda similar - we can await `this.space.syncDone` to achieve the same affect, which is a public but undocumented property. Not ideal, but better than the current solution.

Regarding the timeout, handling that case and throwing an error is simple enough, but we're kinda stuck once the timeout happens. We can't abort the sync, so we can't reliably retry the operation. We'll have to just disable 3Box completely, and not allow the user to enable it (unless they uninstall and reinstall the extension).

If it _does_ return after the timeout, we could allow the user to retry/re-enable it. But if it times out and never returns, we're out of luck.

Gudahtt replied:
So it looks like the changes to `onSyncDone` to return a promise haven't been released yet. I don't think they need to block the release though, as long as we handle the timeout case. The error will just get caught in handling the timeout. Far from ideal, as we can't recover from it and we force the user to wait, but we can improve things whenever those changes are released.

Gudahtt replied:
Resolved for now with [47e63ed](https://github.com/MetaMask/metamask-extension/pull/6972/commits/47e63edf9ceedafa7f6f0b2a432d032297aeef7d). Explained here: https://github.com/MetaMask/metamask-extension/pull/6972#issuecomment-522718296

Gudahtt commented on test/setup.js:
```diff
@@ -5,3 +5,6 @@ require('@babel/register')({
 window.SVGPathElement = window.SVGPathElement || { prototype: {} }
+window.fetch = window.fetch || function fetch () { return Promise.resolve() }
+global.indexedDB = {}
+global.fetch = global.fetch || function fetch () { return Promise.resolve() }
```
It looks like this was accidentally added twice

Gudahtt commented on ui/app/components/app/modals/threebox-restore-confirm/threebox-restore-confirm.component.js:
```diff
@@ -0,0 +1,58 @@
+    } = this.props
+    restoreFromThreeBox(address)
+      .then(() => {
+        setRestoredFromThreeBox(true)
```
Does this need to be here at all? `restoreFromThreeBox` should already set this to true.

danjm replied:
Your right, `setRestoredFromThreeBox(true)` can be removed

Gudahtt commented on ui/app/components/app/modals/threebox-restore-confirm/threebox-restore-confirm.component.js:
```diff
@@ -0,0 +1,58 @@
+    restoreFromThreeBox(address)
+      .then(() => {
+        setRestoredFromThreeBox(true)
+        setThreeBoxSyncingPermission(true)
```
I'm not sure whether this is needed either. Isn't it impossible for `restoreFromThreeBox` to succeed unless 3box was already initialized, and doesn't it being initialized imply that 3box was already allowed?

danjm replied:
Hmm... that looks to be the case as well. What we do need here is to `_registerUpdates()`

danjm replied:
which is wrapped with the "public" method `turnThreeBoxSyncingOn()`

Gudahtt replied:
Ah I see, makes sense.

Gudahtt commented on ui/app/pages/routes/index.js:
```diff
@@ -98,6 +98,27 @@ class Routes extends Component {
+      restoredFromThreeBox,
+    } = this.props
+
+    if (!prevProps.threeBoxSynced && threeBoxSynced && restoredFromThreeBox === null) {
```
I'm not clear on what purpose `restoredFromThreeBox` serves. It looks like the primary purpose is to prevent showing the restore confirm modal more than once in a session. However, the only case where `threeBoxSynced` could conceivably be switching from false to true is for the initial sync, unless 3box was toggled off then on again.

Is there any harm in showing the modal again after 3Box is re-enabled though? It might be useful to give the user some feedback about when syncing has finished. I'm not sure.

danjm replied:
The reason that we don't want to show the modal after 3Box is re-enabled is that we want to prevent actual restoring/overwriting of local data with remote data at all times other then after import. This is to prevent a case such as: local state is A -> threebox is turned off -> local state changes to B -> threebox is turned on -> modal is shown and confirm -> local state is set back to A

We could, however, show a different modal when it is re-enabled / toggled on. Probably not necessary at this time.

danjm replied:
Moreover, `restoredFromThreeBox` persists across sessions, whereas `threeBoxSynced` defaults to false for every new session. So `restoredFromThreeBox` tells us whether the user did in fact approval the overwriting of local data with remote data after an import. This ends up preventing the modal from showing up more than "once per import".

Gudahtt replied:
This all makes sense; I somehow got confused about what happened upon confirmation.

Gudahtt commented on ui/app/pages/routes/index.js:
```diff
@@ -98,6 +98,27 @@ class Routes extends Component {
+          if (lastUpdated) {
+            show3BoxRestoreConfirmModal(lastUpdated)
+          } else {
+            setThreeBoxSyncingPermission(true)
```
Similar to the other case, I don't see how this could reach this state without this permission being `true` already.

danjm replied:
You are right again, and same as before this needs to be replaced with `turnThreeBoxSyncingOn()`

Gudahtt commented on ui/app/pages/routes/index.js:
```diff
@@ -341,6 +362,13 @@ Routes.propTypes = {
   providerId: PropTypes.string,
   providerRequests: PropTypes.array,
   autoLogoutTimeLimit: PropTypes.number,
+  completedOnboarding: PropTypes.bool,
```
It looks as though this prop was added accidentally.

Gudahtt commented on ui/app/ducks/app/app.js:
```diff
@@ -756,6 +758,21 @@ function reduceApp (state, action) {
         loadingMethodData: false,
       })
 
+    case actions.SHOW_SEED_PHRASE_BACKUP_AFTER_ONBOARDING:
```
These two actions look like they may have been added accidentally in resolving a merge conflict.

Gudahtt replied:
Hmm.... actually, `HIDE_SEED_PHRASE_BACKUP_AFTER_ONBOARDING` does appear to be used, as I just discovered.

`SHOW_SEED_PHRASE_BACKUP_AFTER_ONBOARDING` is still unused though.

danjm replied:
No, you were right, both were accidentally re-added with a merge conflict. Here is where they were deliberately removed on a PR that was merged to develop: https://github.com/MetaMask/metamask-extension/pull/6961/commits/58f90d734ba885a2dbdd14f3c0761d0951c6cfd5

`HIDE_SEED_PHRASE_BACKUP_AFTER_ONBOARDING` is dispatched by a method that is never actually called. I will deal with that dead code in a separate PR.

Gudahtt commented on ui/app/store/actions.js:
```diff
@@ -2798,3 +2825,79 @@ function setSeedPhraseBackedUp (seedPhraseBackupState) {
   }
 }
+
+function hideSeedPhraseBackupAfterOnboarding () {
```
It looks like this was added accidentally.

Gudahtt replied:
Oh wait... nevermind. This is needed - it was just missing. That's concerning.

danjm replied:
In fact, it is not needed. The case where it appears in the code is never called. I've removed the addition of this via a rebase, and will remove the case where it is called in a separate PR.

Gudahtt replied:
It looks like this is still here - did it get re-added?

Gudahtt replied:
It looks like this should be here - it's just broken on develop right now :/

Gudahtt commented on ui/app/ducks/app/app.js:
```diff
@@ -756,6 +758,21 @@ function reduceApp (state, action) {
+        showingSeedPhraseBackupAfterOnboarding: false,
+      })
+
+    case actions.SHOW_3BOX_MODAL_AFTER_IMPORT:
```
This action appears to not exist

danjm replied:
another good catch

Gudahtt commented on ui/app/pages/first-time-flow/create-password/import-with-seed-phrase/import-with-seed-phrase.component.js:
```diff
@@ -140,6 +141,8 @@ export default class ImportWithSeedPhrase extends PureComponent {
 
       setSeedPhraseBackedUp(true).then(() => {
+        const address = vault.keyrings[0].accounts[0]
+        initializeThreeBox(address)
```
Maybe this would be better done in `metamask-controller`? I found it a bit hard to keep track of where 3Box was being initialized in each of the three scenarios: unlocking, creating a new account, and restoring an account. The first two cases are in the controller, only the last is in the UI.

Gudahtt commented on ui/app/pages/home/home.component.js:
```diff
@@ -45,6 +45,8 @@ export default class Home extends PureComponent {
     shouldShowSeedPhraseReminder: PropTypes.bool,
     rejectProviderRequestByOrigin: PropTypes.func,
     isPopup: PropTypes.bool,
+    show3BoxModalAfterImport: PropTypes.bool,
```
These props don't appear to be used anymore

danjm replied:
:+1:

danjm committed:
Remove unnecessary property of threebox controller provider

Gudahtt commented on app/scripts/controllers/threebox.js:
```diff
@@ -0,0 +1,181 @@
+  setThreeBoxSyncingPermission (newThreeboxSyncingState) {
+    const currentState = this.store.getState()
+    this.store.updateState({
+      ...currentState,
```
This can be omitted - `updateState` updates top-level properties, leaving the rest of the state unchanged. This pattern of including a spread of the existing object is only needed for updating non-top-level things.

danjm committed:
Remove unnecessary method calls after restoration from 3box in the threebox-restore-confirm modal.

Gudahtt commented on ui/app/store/actions.js:
```diff
@@ -2428,6 +2437,24 @@ function setAutoLogoutTimeLimit (value) {
+  return dispatch => {
+    dispatch(actions.showLoadingIndication())
+    return new Promise((resolve, reject) => {
+      background.setThreeBoxSyncing(newThreeBoxSyncingState, (err) => {
```
`setThreeBoxSyncing` doesn't appear to exist. I encountered an error for this locally by toggling the 3Box setting.

Gudahtt replied:
Similarly, the state `metamask.threeboxSyncing` doesn't exist either

Gudahtt commented:
Are we sure that this won't ever overwrite data? It does sync on unlock, which seems like it could wipe out local state. There's no warning or anything about ti either.

I thought the plan to avoid overwriting data was to only sync on import? Maybe I'm misremembering.

danjm committed:
Replace setThreeBoxSyncingPermission calls in routes/index.js with turnThreeBoxSyncingOn

danjm committed:
Replace erroneous use of  with

danjm committed:
Replace erroneous use of threeboxSyncing with threeBoxSyncingAllowed in advancted-tab directory

danjm committed:
Lint fixes for 3box changes

Gudahtt committed:
Log errors encountered when updating 3Box

Gudahtt committed:
Remove unnecessary parameter from state update

Gudahtt committed:
Add timeout to initial 3Box sync

The initial 3Box sync will now timeout after 1 minute. If the timeout
is triggered, 3Box is disabled and cannot be re-enabled unless the
initial sync does finally finish. If it never finishes, 3Box cannot
be enabled unless the extension is reinstalled.

The Advanced Settings page was updated to show this option as disabled
in that circumstance, with a new discription explaining why it's
disabled. The UI here could certainly be improved.

Additionally, "on" and "off" labels were added to the toggle to match
the other toggles on the Advanced Settings page.

Gudahtt commented:
I've added a timeout for the initial sync with 3Box, after which 3Box remains disabled until the sync finishes (if ever). Here's what the settings page toggle looks like when 3Box is disabled:

![3box_disabled](https://user-images.githubusercontent.com/2459287/63292969-48e2fc00-c29d-11e9-89e1-aac487855cf7.png)

Really not great, as a user probably won't know what "initial sync" means, nor would they know how to fix this. But we're in an awkward spot, as there isn't any obvious way to fix the problem without... waiting, and hoping it fixes itself, or reinstalling the extension.

Gudahtt committed:
Use non-minified 3Box module

We had previously used the minified 3Box module to avoid a build error
encountered when `envify` was processing the `libp2p` module (which is
used by 3Box). The build would fail because `esprima` (used by `envify`)
is incompatible with the object spread/rest operator (which is used in
`libp2p`).

That issue has been solved by adding a global Babelify transformation
specifically for transpiling out the object rest/spread operator from
dependencies. It has been targetted to only affect `libp2p` to avoid
extending the build time too much. This workaround can be used until
a new version of `esprima` is released that includes this bug fix.

danjm committed:
Use app key addresses for threebox

danjm committed:
Replace use of modal for confirming 3box restoration with a home notification

danjm committed:
Adds e2e tests for restoring from threebox

Gudahtt commented on app/scripts/controllers/threebox.js:
```diff
@@ -42,7 +41,9 @@ class ThreeBoxController {
       processPersonalMessage: (msgParams) => {
-        return Promise.resolve(signPersonalMessage(msgParams))
+        return Promise.resolve(keyringController.signPersonalMessage(msgParams, {
+          withAppKeyOrigin: '3box.metamask.io',
```
Shouldn't this be prefixed with `https://`?

danjm replied:
maybe, why do you say so?

Gudahtt replied:
Without the scheme, there's no way to distinguish it from other protocols, so we effectively disallow other protocols.

The scheme was included in @danfinlay 's description of this feature in the EIP thread: https://ethereum-magicians.org/t/eip-erc-app-keys-application-specific-wallet-accounts/2742/3

danfinlay replied:
Yeah, I agree it should have a prefix of some sort. Strangely this is a decent opportunity for a non-existent prefix, to ensure no DNS attack could impersonate this domain. For example: `browser-extension://3box.metamask.io`.

Or, maybe it would be crazy, but we could even pick a domain that is friendly towards portability between wallets. Something like `wallet://backup` or something...

Gudahtt commented on app/_locales/en/messages.json:
```diff
@@ -505,6 +505,12 @@
   "customRPC": {
     "message": "Custom RPC"
   },
+  "dataBackupFound": {
```
This message appears to be unused

danjm committed:
Update eth-keyring-controller to 5.1.0

danjm committed:
Correct parameters passed to getAppKeyAddress in threebox.js

danjm committed:
Add prefix to origin passed to getAppKeyAddress in threebox.js

danjm committed:
Remove unused locale message.

danjm committed:
Prevent CORS errors in firefox e2e tests

danjm committed:
Ensure extraneous scripts are excluded from the local test dev build

Gudahtt commented on gulpfile.js:
```diff
@@ -208,6 +209,47 @@ gulp.task('manifest:testing', function () {
+  'bg-libs.js': true,
+}
+
+gulp.task('manifest:testing-local', function () {
```
Interesting - so this task is only used for `test:copy`, which is only used by `dev:test`, which itself is only called from the `start:test` npm script.

What is the `start:test` script used for anyway? Why does it need the `localhost/8889` permission? It looks like it's basically the same as the dev build except that `IN_TEST` is set to true. I'm not sure why it'd need a special permission.

danjm replied:
`start:test` shares `json.permissions = [...json.permissions, 'webRequestBlocking', 'http://localhost/8889']` with the `manifest:testing` task, which is run with `build:test`. Whenever we are running tests, we need these permissions. `'webRequestBlocking'` is needed for tests so that we can prevent certain network requests. `'http://localhost/8889'` is being added with this PR so that we can make certain http requests to a local server.

Like `start`, `start:test` intentionally excludes `bg-libs.js` from the background files listed in the manifest... because `start` doesn't build that file (I can't remember why that decision was made).

`start:test` itself is useful when you want a watched build of metamask running while working on e2e tests, or if you want to run metamask in the browser with `IN_TEST` set to true

Gudahtt replied:
Right - I had thought at the time that the `localhost` permission was only added in this task - I hadn't realized it was in `manifest:testing` as well. I had thought that was the reason for creating `manifest:testing-local` in the first place.

But rather it looks like the reason was so that `bg-libs` could be excluded.

This seems fine, though I remain of the opinion that we'd be better off splitting those background libraries out in development and testing just as we do in production. Particularly as the gulpfile grows more complex to accommodate this.

Gudahtt commented on gulpfile.js:
```diff
@@ -208,6 +209,47 @@ gulp.task('manifest:testing', function () {
+        ...json.background,
+        scripts: json.background.scripts.filter(scriptName => !scriptsToExcludeFromBackgroundDevBuild[scriptName]),
+      }
+      json.permissions = [...json.permissions, 'webRequestBlocking', 'http://localhost/8889']
```
```suggestion
      json.permissions = [...json.permissions, 'webRequestBlocking', 'http://localhost:8889']
```
I'm guessing that `8889` is a port, not a path?

danjm replied:
So, I don't understand why, but `http://localhost/8889` and `http://localhost/*` work, but `http://localhost:8889`, `http://localhost:*` and `http://localhost` do not

Gudahtt replied:
Interesting - apparently the match pattern here doesn't support ports?
https://bugzilla.mozilla.org/show_bug.cgi?id=1362809

I think we're better off omitting the port rather than including this non-existent "8999" path. I'm also curious whether this works on Firefox

Gudahtt replied:
Thoughts on whether we should change this to `localhost/*`? I remain confused about why `localhost/8999` works in the first place. If you feel inclined to leave it as `localhost/8999` we should at least make sure it works on Firefox, as I can't see any documentation to explain why it _does_ work, so it could be some sort of fluke.

Gudahtt commented on test/e2e/mock-3box/server.js:
```diff
@@ -0,0 +1,37 @@
+    let body = ''
+    request.on('data', chunk => {
+      body += chunk.toString() // convert Buffer to string
+      // console.log('Mock3Box POST body', body)
```
Probably no need to keep this - I assume it was just used during testing.

danjm replied:
fixed in acd4f9bc7

Gudahtt commented on test/e2e/run-all.sh:
```diff
@@ -64,3 +64,12 @@ concurrently --kill-others \
+    'node test/e2e/mock-3box/server.js' \
+    'yarn ganache:start' \
+    'yarn dapp' \
+    'sleep 5 && mocha test/e2e/threebox.spec'
```
Interesting that the end-of-file newline is missing. Our `.editorconfig` file does ask for one - maybe your editor doesn't have the editorconfig plugin installed?

danjm replied:
fixed in acd4f9bc7

Gudahtt commented on test/e2e/threebox.spec.js:
```diff
@@ -0,0 +1,328 @@
+    // are closing any extraneous windows to reset us to a single window before continuing.
+    const [tab1] = await driver.getAllWindowHandles()
+    await closeAllWindowHandlesExcept(driver, [tab1])
+    console.log('tab1', tab1)
```
Another console.log from testing I assume

danjm replied:
fixed in acd4f9bc7

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -1664,6 +1697,10 @@ module.exports = class MetamaskController extends EventEmitter {
   }
 
+  async initializeThreeBox (address) {
+    await this.threeBoxController.new3Box(address, true)
```
Maybe this should always grab the first address, rather than accepting an address as an input. Otherwise it could end up being initialized with different accounts depending on where in the flow it became initialized.

danjm replied:
done in 40a3c5879

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -514,6 +529,14 @@ module.exports = class MetamaskController extends EventEmitter {
+      // 3Box
+      setThreeBoxSyncingPermission: nodeify(threeBoxController.setThreeBoxSyncingPermission, threeBoxController),
+      restoreFromThreeBox: nodeify(threeBoxController.restoreFromThreeBox, threeBoxController),
+      setRestoredFromThreeBox: nodeify(threeBoxController.setRestoredFromThreeBox, threeBoxController),
```
This is a bit of a nitpick, but maybe this method should be changed to no longer accept a `bool` parameter. It is never and should never be called with a `true` parameter, because that happens as part of `restoreFromThreeBox`.

Maybe it could be called `setNotRestoredFromThreeBox` - sounds kinda dumb, but I can't think of a better sounding suggestion at the moment.

danjm replied:
Addressed in 37f6dfb11

danjm committed:
Move threeBoxLastUpdate state from home.component to redux

danjm committed:
Threebox PR code cleanup

danjm committed:
Always use first address when initializing threebox

danjm commented on ui/app/pages/home/home.component.js:
```diff
@@ -59,6 +71,30 @@ export default class Home extends PureComponent {
+      threeBoxLastUpdated,
+    } = this.state
+    if (threeBoxSynced && restoredFromThreeBox === null && threeBoxLastUpdated === null) {
+      getThreeBoxLastUpdated()
```
moved to redux in aaf0f5ba3

danjm commented on ui/app/pages/home/home.component.js:
```diff
@@ -69,8 +105,16 @@ export default class Home extends PureComponent {
+    const {
+      threeBoxLastUpdated,
+    } = this.state
+    console.log('HOME threeBoxLastUpdated', threeBoxLastUpdated)
```
fixed as part of 40a3c5879

danjm committed:
Replace setRestoredFromThreeBox api with setRestoredFromThreeBoxToFalse and setRestoredFromThreeBoxToTrue

Gudahtt commented on ui/app/pages/home/home.container.js:
```diff
@@ -32,11 +40,29 @@ const mapStateToProps = state => {
   unsetMigratedPrivacyMode: () => dispatch(unsetMigratedPrivacyMode()),
+  turnThreeBoxSyncingOn: () => dispatch(turnThreeBoxSyncingOn()),
+  setupThreeBox: () => {
+    dispatch(getThreeBoxLastUpdated())
```
It is a bit awkward that we're storing the lastUpdated time of 3Box in two places - the app state and the metamask state. Moreover, the app state last updated time is only updated once per session, then never kept up-to-date after that.

Granted, it does still serve its purpose perfectly well, but it does seem confusing.

Could we just use `metamask.lastUpdated` directly, get rid of the getter method, and get rid of the app state?

Gudahtt commented on ui/app/store/actions.js:
```diff
@@ -455,20 +463,21 @@ function createNewVaultAndRestore (password, seed) {
       .then(() => {
         dispatch(actions.showAccountsPage())
         dispatch(actions.hideLoadingIndication())
+        return vault
```
Strictly speaking I don't think we need this change anymore, but it doesn't seem like it does any harm.

Gudahtt reviewed (changes requested):
Given that we're now intending to release this as opt-in, we'd might as well make that change here before we merge this.

Maybe we could also merge this into develop sooner rather than later, and hide the option to enable 3Box behind an environment variable? Like a non-user facing feature flag. I don't think we have anything like that currently, but it'd save you the work of keeping this branch up-to-date these next couple weeks.

Gudahtt commented:
Just adding a note here that #7141 conflicts with this PR in a way that git won't detect. It made a change to `development/metamaskbot-build-announce.js` that included a hard-coded list of JavaScript bundles, so that list needs to be updated to rename `lib.js` to `ui-libs.js` and to include `bg-libs.js`

danjm committed:
Update development/metamaskbot-build-announce.js to include ui-libs and bg-libs in hard coded bundle list

danjm committed:
Update test/e2e/threebox.spec.js to use new helpers added with pull #7144

danjm committed:
Make setFeatureFlag available on the ui window during testing

danjm committed:
Hide threebox feature behind a feature flag that can only be activated via dev console

Gudahtt commented on app/scripts/migrations/037.js:
```diff
@@ -0,0 +1,35 @@
+const clone = require('clone')
+
+/**
+ * The purpose of this migration is to remove the {@code privacyMode} feature flag.
```
This comment should be updated

Gudahtt commented on ui/app/store/actions.js:
```diff
@@ -2789,3 +2799,125 @@ function setSeedPhraseBackedUp (seedPhraseBackupState) {
+function initializeThreeBox () {
+  return (dispatch, getState) => {
+    const state = getState()
+    console.log('!!! getFeatureFlags(state)', getFeatureFlags(state))
```
Leftover `console.log` statements here

Gudahtt commented on ui/app/pages/home/home.component.js:
```diff
@@ -125,6 +152,25 @@ export default class Home extends PureComponent {
                       />,
                     },
+                    {
+                      shouldBeRendered: threeBoxLastUpdated && restoredFromThreeBox === null,
```
Conditions like this might leave us in a strange state if the flag is turned on, then off again.

This situation wouldn't occur unless someone enabled it in the dev console of course, but do you think it's worth adding the flag here to prevent it anyway? It's possible an end-user might enable 3Box in the dev console then disable it, and get in a weird state like this. Maybe unlikely though - I'm OK with ignoring this case if you'd prefer. If you did want to handle it, there are a few other examples around like this as well.

danjm replied:
Good point. Done in 4fcc18a77

Gudahtt commented on app/scripts/metamask-controller.js:
```diff
@@ -198,6 +199,16 @@ module.exports = class MetamaskController extends EventEmitter {
 
     this.addressBookController = new AddressBookController(undefined, initState.AddressBookController)
 
+    this.threeBoxController = new ThreeBoxController({
```
I believe this should be behind the feature flag as well, as `init` is called within the controller constructor.

Or maybe the flag should be used _within_ the controller... not sure.

danjm replied:
Did it in the controller here a203d5862
If we did it in the metamask controller, we would have to had some new approach for dynamically instantiating the threebox controller based on a state change.

Gudahtt replied:
Right, good point. In the controller makes much more sense.

Gudahtt commented:
Also it'd be nice to have tests for the migration. I think all of the recent migrations have been tested, so it should be pretty easy to add using a previous one as a template.

danjm committed:
Remove unnecessary migration of threebox feature flag

danjm committed:
Prevent this.init() call in threebox constructor if feature flag is not turned on

danjm committed:
Prevent threebox notification from showing if feature flag is falsy

danjm committed:
http://localhost/8889 -> http://localhost/* in gulp manifest:testing tasks

danjm commented:
@Gudahtt I realized we don't actually need a migration. The feature flag will be `undefined` until it is set to true, which will have the same result as if we migrated to explicitly be `false`.

Gudahtt reviewed (approved):
Looks good!
